### PR TITLE
Allow overriding runAsNonRoot behaviour for proxyInit

### PR DESCRIPTION
**Pull request by alex-berger**
Fixes https://github.com/linkerd/linkerd2/issues/7308

Signed-off-by: Alexander Berger <alex-berger@gmx.ch>

### CODE REVIEW

File: charts/partials/templates/_proxy-init.tpl [hunk 1]
```diff
@@ -46,7 +46,7 @@ securityContext:
     {{- include "partials.proxy-init.capabilities.drop" . | nindent 4 -}}
     {{- end }}
     {{- end }}
-  {{- if .Values.proxyInit.closeWaitTimeoutSecs }}
+  {{- if or .Values.proxyInit.closeWaitTimeoutSecs .Values.proxyInit.runAsRoot }}
```
Comment (mateiidavid) [inline charts/partials/templates/_proxy-init.tpl:49]: I think we need to also change [L32-L34](https://github.com/linkerd/linkerd2/blob/94c04fd06ac62d09b563c3d515580c5328dd9841/charts/partials/templates/_proxy-init.tpl#L32-#L34) here.

When I try to install with `runAsRoot=true`, I get errors for all control plane deployments; `allowPrivilegeEscalation` cannot be set to false when `privileged` is set to true. I think if we add the same conditional to L32 we should be good to go.

Error message, to get this, I built your branch and tried to install linkerd through the CLI (`linkerd install --set "proxyInit.runAsRoot=true" | kubectl apply -f -`)
```
Error from server (Invalid): error when creating "STDIN": Deployment.apps "linkerd-proxy-injector" is invalid: spec.template.spec.initContainers[0].securityContext: Invalid value: core.SecurityContext{Capabilities:(*core.Capabilities)(0xc01a587260), Privileged:(*bool)(0xc02460f5aa), SELinuxOptions:(*core.SELinuxOptions)(nil), WindowsOptions:(*core.WindowsSecurityContextOptions)(nil), RunAsUser:(*int64)(0xc02460f5d0), RunAsGroup:(*int64)(nil), RunAsNonRoot:(*bool)(0xc02460f5ac), ReadOnlyRootFilesystem:(*bool)(0xc02460f5ab), AllowPrivilegeEscalation:(*bool)(0xc02460f57f), ProcMount:(*core.ProcMountType)(nil), SeccompProfile:(*core.SeccompProfile)(nil)}: cannot set `allowPrivilegeEscalation` to false and `privileged` to true
```

**Review by mateiidavid — COMMENTED** [review]
@alex-berger thank you for the PR!

Left a comment, I think after we get that out of the way this should be good to go 🙌🏻

File: charts/partials/templates/_proxy-init.tpl (hunk 1 above)
Comment (alex-berger) [reply charts/partials/templates/_proxy-init.tpl:44]: OK, I fixed that and also collapsed the two `if` statements into one to simplify things.

**Review by alpeb — APPROVED** [review]
Thanks @alex-berger , this looks good to me :+1: 
A nice followup to this would be to add a check to `linkerd check` that identifies if the node runtime container is not `containerd`. The node's status has a [field](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#nodesysteminfo-v1-core) `nodeInfo.containerRuntimeVersion` that can be leveraged for this purpose. Would you be up to take that up? :wink:

**Review by mateiidavid — APPROVED** [review]
🚢

**Pull request merged**